pkg/services/govmomi/service.go

```diff
@@ -250,13 +263,176 @@ func (vms *VMService) reconcileNetworkStatus(ctx *virtualMachineContext) error {
+				return false, err
+			}
+			if apierrors.IsNotFound(err) {
+				ctx.Logger.Info(fmt.Sprintf("creating IPAddressClaim %s", ipAddrClaimName))
+				claim := &ipamv1.IPAddressClaim{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      ipAddrClaimName,
+						Namespace: ctx.VSphereVM.Namespace,
+						OwnerReferences: []metav1.OwnerReference{
+							{
+								APIVersion: ctx.VSphereVM.APIVersion,
+								Kind:       ctx.VSphereVM.Kind,
+								Name:       ctx.VSphereVM.Name,
+								UID:        ctx.VSphereVM.UID,
+							},
+						},
+						Finalizers: []string{infrav1.IPAddressClaimFinalizer},
+					},
+					Spec: ipamv1.IPAddressClaimSpec{PoolRef: poolRef},
+				}
+				if err = ctx.Client.Create(ctx, claim); err != nil {
+					return false, err
+				}
+				msg := "Waiting for IPAddressClaim to have an IPAddress bound"
+				markIPAddressClaimedConditionWaitingForClaimAddress(ctx.VSphereVM, msg)
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func (vms *VMService) reconcileIPAddresses(ctx *virtualMachineContext) (bool, error) {
+	ctx.IPAMState = map[string]infrav1.NetworkDeviceSpec{}
+	for devIdx, device := range ctx.VSphereVM.Spec.Network.Devices {
+		var ipAddrs []string
+		var gateway4 string
+		var gateway6 string
+
+		for poolRefIdx := range device.AddressesFromPools {
+			// check if claim exists
+			ipAddrClaim := &ipamv1.IPAddressClaim{}
+			ipAddrClaimName := fmt.Sprintf("%s-%d-%d", ctx.VSphereVM.Name, devIdx, poolRefIdx)
+			ipAddrClaimKey := apitypes.NamespacedName{
+				Namespace: ctx.VSphereVM.Namespace,
+				Name:      ipAddrClaimName,
+			}
+			var err error
+			ctx.Logger.Info(fmt.Sprintf("fetching IPAddressClaim %s with key %s", ipAddrClaimName, ipAddrClaimKey.String()))
+			if err = ctx.Client.Get(ctx, ipAddrClaimKey, ipAddrClaim); err != nil && !apierrors.IsNotFound(err) {
+				ctx.Logger.Info(fmt.Sprintf("error fetching IPAddressClaim %s: %s", ipAddrClaimName, err.Error()))
+				return false, err
+			}
+
+			ipAddrName := ipAddrClaim.Status.AddressRef.Name
+			ctx.Logger.Info(fmt.Sprintf("fetched IPAddressClaim %s with AddressRef.Name %s", ipAddrClaimName, ipAddrName))
+			if ipAddrName == "" {
+				ctx.Logger.Info(fmt.Sprintf("name was empty on claim IPAddressClaim.Status.AddressRef.Name %s: %s", ipAddrClaimName, ipAddrName))
+				msg := "Waiting for IPAddressClaim to have an IPAddress bound"
+				markIPAddressClaimedConditionWaitingForClaimAddress(ctx.VSphereVM, msg)
+				return false, errors.New(msg)
+			}
+
+			ipAddr := &ipamv1.IPAddress{}
+			ipAddrKey := apitypes.NamespacedName{
+				Namespace: ctx.VSphereVM.Namespace,
+				Name:      ipAddrName,
+			}
+			if err = ctx.Client.Get(ctx, ipAddrKey, ipAddr); err != nil {
+				return false, err
+			}
+
+			toAdd := fmt.Sprintf("%s/%d", ipAddr.Spec.Address, ipAddr.Spec.Prefix)
+			parsedIP, _, err := net.ParseCIDR(toAdd)
+			if err != nil {
+				msg := fmt.Sprintf("IPAddress %s/%s has invalid ip address: %q",
```

Comment:
We had been logging errors, but removed them because we were seeing duplicate logs. 
We're seeing that errors returned during reconciliation are logged. Are there cases where the errors would not be logged?